python/mxnet/amp/amp.py

```diff
@@ -0,0 +1,262 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# coding: utf-8
+"""Functions for enabling AMP (automatic mixed precision)."""
+__all__ = ['init', 'init_trainer', 'scale_loss']
+
+from types import MethodType
+import logging
+import contextlib
+import numpy as np
+
+from .. import symbol
+from ..symbol import Symbol
+from ..symbol import contrib as symbol_contrib
+from .. import ndarray
+from ..ndarray import NDArray
+from ..ndarray import contrib as ndarray_contrib
+from . import lists
+from ..gluon import trainer
+from .. import optimizer as opt
+from .loss_scaler import LossScaler
+
+def _cast_symbol_NDArray(s, dtype):
+    if isinstance(s, Symbol):
+        return symbol.amp_cast(s, dtype=dtype)
+    elif isinstance(s, NDArray):
+        if (s.dtype != dtype and
+                (s.dtype == np.float16 or s.dtype == np.float32) and
+                s.context.device_type != 'cpu'):
+            return ndarray.amp_cast(s, dtype=dtype)
+        else:
+            return s
+    else:
+        return s
+
+def _wrap_symbol_functions(module):
+    def _ndarray_wrapper(f, target_dtype, cond_arg=None):
+        def _new_fun(*args, **kwargs):
+            if cond_arg is not None:
+                if (cond_arg[0] not in kwargs or
+                        kwargs[cond_arg[0]] not in cond_arg[1]):
+                    return f(*args, **kwargs)
+            new_args = list(map(lambda x: _cast_symbol_NDArray(x, target_dtype), args))
+            args = tuple(new_args)
+            kwargs = {k: _cast_symbol_NDArray(v, target_dtype) for k, v in kwargs.items()}
+            return f(*args, **kwargs)
+        _new_fun.__name__ = f.__name__
+        _new_fun.__module__ = f.__module__
+        _new_fun.__doc__ = f.__doc__
+        return _new_fun
+
+    def _symbol_wrapper(f, target_dtype, cond_arg=None):
```

Comment:
What is the expected workflow for deploying models trained with AMP? 

FP16 is not well supported on CPU and I assume we do not want to include amp cast/multi-cast in the exported symbol.json file. For users who want to do deployment on GPUs with fp16 support with c++, exporting the symbols with casts is fine. For users who want to deploy models on CPU, are they expected to export only when `amp.init()` is not called?

Reply:
All the weights are fp32 and the model itself does not know anything about fp16, so user can import and export fp32 models.

Reply:
Hmmm, actually I would need to double check it. If you just save parameters from the Gluon model then they all will be in fp32, but I'm not sure about saving the entire symbol from the network built while AMP was active.

Reply:
The workaround of saving params only and then building the network again without AMP with those parameters and exporting symbol would work, but it is definitely not ideal.

Reply:
That's right. Requiring exporting without AMP does not sound ideal. The best way is to do this kind of graph rewriting in the backend and export the original one

Reply:
Added a graph pass that eliminates amp casts when saving the model.